### ISSUE TITLE
Parallelization, or pickle save graph.MCP_Geometric fails

### ISSUE THREAD

**Issue by ozak** [scikit-image/scikit-image#1353]
Hi all,

Has anyone successfully tried to parallelize the object generated by graph.MCP_Geometric? Or is there a way to save the Object and reuse it in other computations?

I am trying to run some optimal path calculations using a cluster with parallel IPython. My non-parallelized code works perfectly fine on a machine with 64GB RAM. The problem is that the nodes on the cluster only have 23GB RAM. My data is a ~3GB Geotiff file, so when I execute

```
mcp = graph.MCP_geometric(geotiff)
```

memory usage at some point increases over 25GB, and ends up with a usage of about 15GB. For this reason I parallelized using IPython as follows

```
from __future__ import division
from IPython.parallel import Client
# Setup the clients, direct views, and balanced views
c = Client()
c.ids
dview = c[:]
view = c.load_balanced_view()
dview.activate()

%%px --local
"""docstring for costsnx"""
import sys, os, time
from osgeo import gdal, gdalnumeric, ogr, osr
import numpy as np
import pandas as pd
import skimage.graph as graph

# Import large raster 
raster = './large.tif'
costs = gdalnumeric.LoadFile(raster)

# Create mcp object
@dview.parallel(block=True)
def costnx(costs):
    # Create Cost surface to work on
    mcp=graph.MCP_Geometric(costs)
    return mcp
```

Although it supposedly works, since there is no error message, trying to use the mcp object for computation fails. In particular, issuing something like

```
cumulative_costs, traceback = mcp.find_costs([[10,10]])
```

fails. Another solution I considered was saving the mcp object as a pickle after processing in the machine with larger memory in order to load the pickle in the cluster. But Pickle fails.

Any ideas?

**Comment by JDWarner** [comment]
I've heard that Dill will serialize almost any Python object, handling many cases where Pickle fails.

Give that a shot first and let us know how it goes - it would be by far the easiest solution.

**Comment by ozak** [comment]
I tried to dump the mcp object with dill and reloading it in a clean session and I get the same error as before, namely

    ---------------------------------------------------------------------------
    TypeError                                 Traceback (most recent call last)
    <ipython-input-10-9ca6b4b9eba0> in <module>()
    ----> 1 mcp.find_costs([[ 5807, 26337] ])

    /usr/local/lib/python2.7/site-packages/skimage/graph/_mcp.so in skimage.graph._mcp.MCP.find_costs (/private/var/folders/z9/fsptst_d35z1s_8c9t467dlw0000gp/T/pip-0JypHD-build/skimage/graph/_mcp.c:6916)()

    /usr/local/lib/python2.7/site-packages/skimage/graph/_mcp.so in skimage.graph._mcp._normalize_indices (/private/var/folders/z9/fsptst_d35z1s_8c9t467dlw0000gp/T/pip-0JypHD-build/skimage/graph/_mcp.c:4456)()

    TypeError: object of type 'NoneType' has no len()

I will try it in IPython paralllel and let you know how it goes there.

**Comment by ozak** [comment]
In IPython it works partially, in the sense that e.g.

    @dview.parallel(block=True)
    def cost(raster):
        # Create Cost surface to work on
        mcp=graph.MCP_Geometric(raster)
        cumulative_costs, traceback = mcp.find_costs([[0,0]])
        cumulative_costs = np.where(cumulative_costs<np.inf, cumulative_costs, np.nan)
        return cumulative_costs.mean()

returns a mean value. This didn't happen without the ``rc[:].use_dill()`` command

    from __future__ import division
    from IPython.parallel import Client
    # Setup the clients, direct views, adn balanced views
    rc = Client(profile='mpi')
    rc.ids
    rc[:].use_dill()
    dview = rc[:]
    view = rc.load_balanced_view()
    dview.activate()

So if one has a small raster that does not get split into parts, one could in parallel find the distances to many points.

Still, the computation is wrong for what I am trying to do, because it split the original raster in various pieces and then created an mcp object for each and computed the distances to point [0,0] on each subraster. But what I am trying to compute is the distance in the full raster. 

Any other ideas?

**Comment by blink1073** [comment]
[We can pickle that](https://m.youtube.com/watch?v=yYey8ntlK_E)

**Comment by ozak** [comment]
@blink1073 can you?

**Comment by blink1073** [comment]
Sorry @ozak, that was me being cheeky.  You may want to try creating a `__reduce__` [method](https://docs.python.org/2/library/pickle.html#pickling-and-unpickling-extension-types).

**Comment by ozak** [comment]
@blink1073 I will check it out. I will also try to understand better the code behind MCP...perhaps one could have it export the graph in some kind of dict or so. Or if one could use MPI+CPython for the code?

**Comment by blink1073** [comment]
If you can capture the state as a dict, pickle that, unpickle it and pass it to the constructor, that is basically what `__reduce__` does.

**Comment by blink1073** [comment]
Make sure to use protocol=-1 for binary pickle.

**Comment by ozak** [comment]
Hmmm...I think that I need to know more about what you are talking about in order to implement it. How do I capture the state? By creating the __reduce__ option? Sorry I am still a newbie.

**Comment by blink1073** [comment]
Sorry, I'm on my phone and drifting off to sleep.  I'll have to get back to you tomorrow.

**Comment by ozak** [comment]
Thanks! I'd really appreciate your help.

**Comment by blink1073** [comment]
@ozak, my impression is that this class is not amenable to pickling because so much is happening in the init function, and it looks like that's where your memory is being chewed up.   Here's my take on a workaround:
Create the object on the main machine, and then use `pickle.dump(mcp.__dict__, protocol=-1)`. 
On other machine, create an `MCP_geometric` with smaller dataset.
Load your pickle file using `pickle.load()`.
Use this loop:

```python
for (key, value) in data.items():
     setattr(mcp, key, value)
```

**Comment by ozak** [comment]
@blink1073 the problem is that this object does not have a ``__dict__`` attribute.

    AttributeError: 'skimage.graph._mcp.MCP_Geometric' object has no attribute '__dict__'

**Comment by blink1073** [comment]
Shucks, I was afraid of that.  It seems like the only option is to manually grab the attributes and store/set them.  New try:

```python
import inspect
variables = [i for i in dir(mcp) if not inspect.ismethod(i)]
data = {}
for var in variables:
    data[var] = mcp[var]
pickle.dump(fid, data, protocol=-1)

...
data = pickle.load(fid)
for (key, value) in data.items():
     setattr(mcp, key, value)
```

**Comment by ozak** [comment]
@blink1073 thanks for the suggestions....it seems this method and objects are not very friendly. Now, I got a 

   TypeError: 'skimage.graph._mcp.MCP_Geometric' object has no attribute '__getitem__'

Issuing ``dir(mcp)`` generates

    ['__class__', '__delattr__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__new__',
     '__pyx_vtable__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__',
     '__subclasshook__', '_reset', 'find_costs', 'goal_reached', 'traceback']

**Comment by blink1073** [comment]
Argh, Cython is not very storage-friendly in general...this [page](http://docs.cython.org/src/tutorial/cdef_classes.html) says "Attributes are by default only accessible from Cython (typed access)".   I think we'd have to add a `cpdef __getstate__(self)` method that returns a dictionary with the current state, and a corresponding `cpdef __setstate__(self, state)`.  Does that sound reasonable @JDWarner?

**Comment by stefanv** [comment]
IIRC, you can also register these methods of picking and unpickling a
specific object outside of the object itself. I don't think we want to
build this into skimage.

**Comment by ozak** [comment]
Can you guide me on how to do this, and I will fork it and try out there without you having to include it in the main skimage?

**Comment by blink1073** [comment]
@stefanv, we can't do that in this case, because the attributes are not accessible outside of C.

@ozak,  here's what it should look like:

```python
cpdef __getstate__(self):
    attrs = ['flat_costs', 'flat_cumulative_costs']   # add the rest here
    state = {}    
    for attr in attrs:
        state[attr] = getattr(self, attr)
    return state

cpdef __setstate__(self, state):
    for (key, value) in state.iteritems():
        setattr(self, key, value)
```

**Comment by blink1073** [comment]
@ozak, I can help you work through this and I think it would be reasonable to include in main skimage.

**Comment by blink1073** [comment]
Come to think of it `getattr/setattr` may not work either.  We may need to make *all* of the attributes visible outside of C using `cdef public` [declarations](http://docs.cython.org/src/tutorial/cdef_classes.html).

**Comment by stefanv** [comment]
What is the advantage of pickling this structure?  As I mentioned in
reply to the original post, this is not a viable approach to
parallelizing the problem.

**Comment by stefanv** [comment]
Hi Omer

I managed to locate my original response.  Here it is:

On 2015-01-19 13:16:31, OOM <omer.ozak@gmail.com> wrote:
> I am trying IPython parallel since I think it should allow me to overcome
> the lack of memory in each node. E.g. If I define a parallel function like
>
> @dview.parallel(block=True)
> def costsnx(costs):
>     mcp=graph.MCP_Geometric(costs)
>     return mcp

I don't think this approach will work.  The MCP implementation is in
Cython, and IPython would not know how to split it up.  Also, it is not
trivially parallelizeable, so you can't split the input and have it work
out of the box.

That said, a straight-forward dynamic programming approach, such as the
one in ``graph.shortest_path`` uses approximately the same amount of
memory as the input, so may work on a single machine in your case.

It is possible to parallelize dynamic programming problems, but
typically you need lightweight threads--the parallelization gives you an
advantage with multiple cores, but is not that useful if you want to
spread memory use across multiple machines where the cost of launching
tasks is high.

For that approach, you may find some of these references interesting:

http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.301.6093
http://research.microsoft.com/apps/pubs/default.aspx?id=208241
http://www.shodor.org/petascale/materials/UPModules/dynamicProgrammingPartI/

Regards
Stéfan

**Comment by ozak** [comment]
Thanks Stéfan! I will try to check it out once I have some time.
